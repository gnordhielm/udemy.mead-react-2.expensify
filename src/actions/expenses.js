import db from '../firebase/firebase'
import uuid from 'uuid'

// Expenses actions

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData={}) => {

  return dispatch => {

    const {
      description='',
      note='',
      amount=0,
      createdAt=0
    } = expenseData

    const expense = {
      description, note, amount, createdAt
    }

    return db.ref('expenses')
      .push(expense)
      .then(ref => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })

  }
}

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})


export const updateExpense = ({ id, changes }) => ({
  type: 'UPDATE_EXPENSE',
  id, changes
})

export const setExpenses = (expenses=[]) => ({
  type: 'SET_EXPENSES', expenses
})

export const startSetExpenses = () => dispatch => {
  return db.ref('expenses')
    .once('value')
    .then(snap => {
      const expenses = []
      snap.forEach(childSnap => {
        expenses.push({
          id: childSnap.key,
          ...childSnap.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
}
