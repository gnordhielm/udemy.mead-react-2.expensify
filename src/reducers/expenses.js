
const initialState = []

export default (state=initialState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ]

    case 'REMOVE_EXPENSE':
      if (!action.id)
        throw new Error('remove expense requires an id')
      return state.filter(({ id }) => id !== action.id)

    case 'UPDATE_EXPENSE':
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.changes
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}
