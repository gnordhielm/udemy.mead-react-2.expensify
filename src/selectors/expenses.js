import moment from 'moment'

const sortByLookup = {
  'date': 'createdAt',
  'amount': 'amount'
}

export const getVisibleExpenses = (expenses, {
  text, sortBy, startDate, endDate
}) => expenses
  .filter(({ description, createdAt }) => {

    const createdAtMoment = moment(createdAt)

    const startDateMatch = startDate ?
      startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ?
      endDate.isSameOrAfter(createdAtMoment, 'day') : true

    const textMatch = description
      .toLowerCase()
      .includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  })
  .sort((a, b) => b[sortByLookup[sortBy]] - a[sortByLookup[sortBy]])

export const getExpensesTotal = expenses => expenses
    .reduce((acc, expense) => acc + expense.amount, 0)
