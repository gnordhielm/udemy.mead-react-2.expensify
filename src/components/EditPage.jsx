import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm.jsx'
import { updateExpense, removeExpense } from '../actions/expenses'

export class EditPage extends React.Component {

  handleSubmit = expense => {
    this.props.updateExpense(expense)
    this.props.history.push('/')
  }

  handleRemove = () => {
    this.props.removeExpense(this.props.expense.id)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <p>Edit expense {this.props.expense.id}</p>
        <button
          onClick={this.handleRemove}
        >Remove</button>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.handleSubmit}
        />
      </div>
    )
  }
}

const mapState = (state, props) => ({
  expense: state.expenses
    .find(({ id }) => id === props.match.params.id)
})

const mapDispatch = dispatch => ({
  updateExpense: expense => dispatch(updateExpense({
    id: expense.id,
    changes: expense
  })),
  removeExpense: id => dispatch(removeExpense({ id }))
})

export default connect(mapState, mapDispatch)(EditPage)
