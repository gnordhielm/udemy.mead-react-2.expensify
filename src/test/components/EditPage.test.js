import 'react-dates/initialize'

import React from 'react'
import { shallow } from 'enzyme'
import { EditPage } from '../../components/EditPage.jsx'
import expenses from '../fixtures/expenses'

let removeExpense, updateExpense, history, wrapper

beforeEach(() => {
  removeExpense = jest.fn()
  updateExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<EditPage
    expense={expenses[0]}
    history={history}
    removeExpense={removeExpense}
    updateExpense={updateExpense}
  />)
})

test('should render EditPage', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle edit expense', () => {

  const editedExpense = expenses[1]

  wrapper.find('ExpenseForm').prop('onSubmit')(editedExpense)

  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(updateExpense).toHaveBeenLastCalledWith(editedExpense)

  expect(wrapper).toMatchSnapshot()

})

test('should handle remove expense', () => {

  const id = expenses[0].id

  wrapper.find('button').simulate('click')

  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(removeExpense).toHaveBeenLastCalledWith(id)

  expect(wrapper).toMatchSnapshot()

})
