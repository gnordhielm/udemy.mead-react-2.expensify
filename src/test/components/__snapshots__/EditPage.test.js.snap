// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle edit expense 1`] = `
<div>
  <p>
    Edit expense 
    abc
  </p>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "abc",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;

exports[`should handle remove expense 1`] = `
<div>
  <p>
    Edit expense 
    abc
  </p>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "abc",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render EditPage 1`] = `
<div>
  <p>
    Edit expense 
    abc
  </p>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "abc",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
